### ISSUE TITLE
FreeType rasterization doesn't work, and has been made mandatory

### ISSUE THREAD

**Issue by mskala** [fontforge/fontforge#685]
This is a priority issue for me because it's blocking the Tsukurimashou 0.8 release, planned for tomorrow.  Encountered in the latest git master version of FontForge, under 64-bit Linux.

Same problem I reported two years ago to the mailing list (http://sourceforge.net/mailarchive/forum.php?thread_name=20111125.100615.444195158.wl%40gnu.org&forum_name=fontforge-devel ):  when FontForge uses FreeType to generate bitmaps, it sometimes produces complete garbage.  The bitmaps, if saved to BDF, are mostly empty, with just a few containing large irregular black areas.  If I generate a bitmap strike in the GUI and then try to look at it in the bitmap window, it sometimes segfaults, suggesting in-memory data structures contain garbage.

The file I posted for that bug report is no longer online, but I've posted a fresh one at https://github.com/fontforge/debugfonts/blob/master/jieub-bt-ps.sfd

In the GUI, there is an option during bitmap generation to use FreeType or not; but native scripting doesn't offer such a choice and will always attempt to use FreeType.

Werner Lemberg found and fixed an unrelated bug in FreeType in response to my report (see the mailing list thread) - apparently FreeType couldn't *read* BDF files with large indices.  But FontForge does not use FreeType for BDF files; that bug is unrelated to mine and only happened to also be noticeable from the test case I provided.  He disclaimed responsibility for my actual bug, which is specific to the case where FontForge generates a bitmap using FreeType.

In 2011, a workaround for me was to build FontForge without FreeType.  But pull #157 made that impossible.

At this point it looks to me like my best option may be to dig into the FontForge source and modify the hardcoded "use FreeType" setting to "no,"  because I need this to work *now*.  But in the longer term, we should figure out what's actually wrong with the FreeType rasterization and fix it, or convince Werner that it is his business to do so.

**Comment by davelab6** [comment]
@mskala still an issue?

**Comment by mskala** [comment]
Yes, it still fails with the current git master revision.

**Comment by frank-trampe** [comment]
Same backtrace?

**Comment by mskala** [comment]
I'm not sure where you want a backtrace from.  The segfault has been fixed since the merge of d556150 in September 2013, but it wasn't the original problem.  The original problem, which remains, is that FontForge produces corrupt bitmaps without crashing.

**Comment by adrientetar** [comment]
Does FreeType provide any benefit in the generation of bitmap? If not we could just disable it by default with a separate config variable on the code path that generates the bitmaps.

**Comment by adrientetar** [comment]
Is there any other issue that prevents you from using the mainline FontForge? I see you have 19 of them opened.

**Comment by mskala** [comment]
I'm not currently entirely prevented from using mainline FontForge, and I do use mainline FontForge for GUI font viewing.  I use FontAnvil for scripting Tsukurimashou builds, and I'm unlikely to switch back.  The reason has more to do with strategic direction than specific bugs:  FontForge's priority is GUI and Python, both of which have negative, not just zero, value in my project's use case, so FontForge isn't the right tool for that job.  Both FontForge and FontAnvil share this bug (#685), and I'm using the 1000 glyphs at a time workaround described above.

I'm not sure that disabling Freetype rasterization is really the best thing to do.  When Freetype works, it does produce nicer-looking results.

**Comment by frank-trampe** [comment]
This may be feasible. What we can do is find where FontForge is calling FreeType functions in this workflow and then check the arguments and their referenced data against the specifications provided by the FreeType documentation. Any volunteers?

**Comment by mskala** [comment]
When FontForge is going to do bitmap generation using Freetype, it creates a temporary Postscript file containing just the glyphs that will be rasterized, writes that to disk in a file created by tmpfile(), mmap()s it, and then calls Freetype to read the resulting chunk of memory.  I modified my copy to use a known filename instead of tmpfile(), and captured the resulting Postscript font file.  It looks okay to me.  I'm now suspicious of Freetype's handing of large files, because the problem specifically appears when FontForge invokes FreeType on a large mmap()ped file, and not when processing identical font data in a smaller file.  I did this on a Mac, by the way, so we can add that to the list of systems that show the same failure.

Is there any code other than FontForge that generates bitmaps using Freetype from an mmap()ped Postscript file, so we could compare against it?

Has anyone other than me attempted to reproduce this bug?

**Comment by frank-trampe** [comment]
@lemzwerg, any ideas?

**Comment by lemzwerg** [comment]
Unfortunately no.  There are no known problem in this area.  BTW, what exactly is `large' for the mmap()ed file?

Do you get a good results if you build FreeType with configure option `--disable-mmap`?

**Comment by mskala** [comment]
My test file is 13 megabytes.  I have seen it happen on slightly smaller files; I'm not sure what is the smallest.  The amount of corruption (percentage of the glyphs that get corrupted) seems to increase with filesize.

FreeType 2.5.5 compiled with --disable-mmap makes no difference.  I am not convinced that that option really has any effect on FreeType compilation.  Recursive grep finds only a couple of calls to mmap() in the entire FreeType code base, they do not seem to be conditional on any configuration setting, and although the configure script parses this option, it doesn't seem to record the value anywhere.  It looks like a do-nothing option to me.  Anyway, even if --disable-mmap has some effect on FreeType by disabling the mmap() calls through a non-obvious mechanism, I don't think that FreeType is calling mmap() in the failing code path.  Rather FontForge calls mmap() and then passes the resulting chunk of memory into FT_New_Memory_Face.  It's not obvious to me why FontForge needs to use mmap() for this; why not just allocate a chunk of memory and read the file into it, or better yet not have a file at all?  I wonder if there is some weirdness involving the use of mmap() on a file from tmpfile(), but the fact that this has shown up identically on both MacOS and Linux makes operating system weirdness less likely.  I will try changing FontForge to read the file into plain ordinary non-mmap()ped memory at some point, but not tonight.  It's close to midnight here in Copenhagen and I should be asleep.

Has anyone other than me attempted to reproduce this bug?

**Comment by lemzwerg** [comment]
There *is* an effect!  If you use `--disable-mmap`, FreeType gets built using the source file `src/base/ftsystem.c` instead of `builds/unix/ftsystem.c`, and only the latter contains calls to `mmap`.  The relevant code can be found in `builds/unix/configure.raw`.

Have you tried to render the temporary PS font with one of the FreeType demo programs?  For example, `ftview` has an command line option `-p` to simulate memory-mapping.

**Comment by mskala** [comment]
The font appears corrupted when loaded with ftview, with or without the -p option.  It may well really be corrupted.  FontForge loads it with no trouble, provided one does not use FreeType in FontForge to rasterize it.  kfontview displays the file with corruption, but I think it's using FreeType.  I don't seem to have a viewer handy that can read .pfb files without using FreeType.  XeLaTeX, when told to use this file as the font for a document, gives the message "Error 11 (driver return code) generating output" and generates an empty (zero-length) PDF file.  I think that means the XeTeX PDF generator, which is a separate process, is segfaulting.  So for the moment I'm guessing that FontForge may be writing a bad file.

The font file is at http://ansuz.sooke.bc.ca/temporary/bmtmp.pfb - that particular version created by FontAnvil because it's what I have on this Mac, but the relevant code is the same and the one from FontForge shows the same behaviour.

**Comment by lemzwerg** [comment]
Thanks for the PFB file.  It is indeed invalid: There are 107791 subroutines!  Within Type 1 charstrings, all numbers must be in the range [-32000,32000] (except arguments to the `div` operator, which in turn must produce a result in this range).  In particular, FreeType can't handle larger values, since it stores all stack values in 16.16 format – or rather, it converts large integer values to the 16.16 format (with a warning), which is of course complete nonsense if the value is used as a subroutine index, thus the distorted rasterization.

*Theoretically*, I could change FreeType to allow larger values, however, this would be a major rewrite of the Type 1 parser, and I strongly doubt whether this is of any value given that it is against the specification.

Even if you restrict yourself to 1000 glyphs per run, there is no guarantee in current FontForge that it doesn't create a font with more than 32000 subroutines...

In summary: FontForge's current route using a PFB for rasterization might lead to invalid results.  No idea how complicated it is to rewrite the stuff so that FontForge tries to split the data into smaller sub-PFBs.

**Comment by mskala** [comment]
I wonder why FontForge is generating so many subroutines.  There are only about 14000 glyphs in the font, so that's an average of more than seven subroutines per glyph.  Most likely, a lot of them are duplicates.

I don't know what the right answer is.  FreeType seems to need a font file (i.e. a sequence of bytes in Postscript or some other standard format) in order to rasterize at all.  FontForge handles that in a very messy way by saving the font into a Postscript file (in principle, on disk) and then loading it with FreeType.  With some minor internal changes, FontForge could and probably should write the temporary file into an in-memory buffer and have FreeType read that, instead of really going through the OS's file interface.  Doing it in smaller batches of glyphs would be a more significant internal change.

**Comment by mskala** [comment]
Okay, on further examination of the font, it looks like those subroutines are associated with hint replacement.  There is some duplication, but even with the duplicates removed, there are almost 82000 distinct hint-replacement subroutines.  And the standard as I understand it requires each of these to really be a subroutine; it's not possible to do hint replacement in a fully standard way with the data stored in any other form.  But maybe it's not necessary to do hint replacement at all, and maybe FreeType can do hint replacement with the data stored in a way that is not fully downward-compatible with the oldest Postscript interpreters.  Failing those possibilities, it may be possible to merge very similar hint-replacement subroutines with very little loss of information until there are few enough to fit in the file format.

**Comment by lemzwerg** [comment]
Maybe FontForge can create an intermediate CFF instead of a PFB?

**Comment by mskala** [comment]
That looks like it may well be possible.  I'll investigate it.  Would CFF make this a non-issue because of being able to represent larger numbers?

**Comment by lemzwerg** [comment]
It's a non-issue in CFF because you don't need hint replacement subroutines at all.  Instead, there is the `hintmask` operator (and `cntrmask`, for which there is no equivalent in Type 1 fonts).

The availability of large integers is another bonus but not important here, I guess.

**Comment by mskala** [comment]
I'm not sure that FontForge decides how many subroutines it needs at the time it saves a file; it looks to me like there may be an array of subroutines generated at some earlier point, which are then dumped out in a fairly dumb way at file-generation time - so that it might still attempt to write a large number of subroutines even if they aren't needed in CFF.  But maybe that array is only "extra" subroutines and wouldn't apply to the hint-replacement ones.  Anyway, I'll look into it.  Thanks for the assistance.

**Comment by KrasnayaPloshchad** [comment]
I found [Wikipedia article](http://en.wikipedia.org/wiki/Freetype) explained FreeType’s subpixel rendering implementation is currently disabled by default to avoid Microsoft's patent about a specific color filtering algorithm for use in LCD display subpixel rendering.

**Comment by lemzwerg** [comment]
While the statement itself is true, it's not clear to me what the last comment wants to tell us...

**Comment by mskala** [comment]
Maybe an attempt to help based on reading the title, and not the content, of this issue?

Anyway, I've gone through the code and found that in principle, making FontForge use CFF for its temporary file would work.  It wouldn't be completely trivial because when FontForge goes to write its PostScript temporary file, it passes a parameter into a function saying "use Postscript"; and just changing that to say "use CFF" instead will cause the CFF write to silently fail, at which point FontForge will fall back to using Postscript again but without any hints - which succeeds, but doesn't process the hints.  To get it right, correct rendering with hints, it needs a little more work to make it *really* write CFF.  But that should be possible.

I got distracted looking at FontForge's handling of temporary files.  Freetype requires a file image, that is, a sequence of bytes in the format that would exist in a font file on disk, but Freetype doesn't require that that actually be a file on disk.  FontForge actually writes a file on disk (which we can hope the OS will keep in cache instead of REALLY on disk), then mmap()s it, and passes the block of memory to Freetype.  That's pretty silly.  I would really like to change FontForge to just write its data into memory.  But that's not trivial because FontForge splits its font-writing code over many different functions, all passing FILE *s around, and it does a fair bit of seeking, opening and closing additional temporary files, and so.  This is a common pattern throughout FontForge's code base.  I wish it wasn't written that way.

Some parts of FontForge's file-writing code actually already include an abstraction layer to allow it to write through a filter (for compression, for example) instead of passing around a FILE * directly, but this is applied inconsistently.

At the very least, FontForge could read its temporary file back into memory before invoking Freetype, and have one less ball in the air at that point.  There's no sensible reason for mmap() to be involved.  Fixing the situation really properly means doing at least one of:

* Expanding the scope of FontForge's current file-writing abstraction layer to cover all file writing
* Writing a new abstraction layer for "things that behave sort of like files but can be stored in RAM" and having FontForge use pointers to such things everywhere it currently uses FILE *s
* Using fopencookie(), which only works on Linux, or some equivalent specific to some other system, to allow FontForge to continue using FILE *s while having them not really be files
* Rewriting it all in C++

At the moment I'm leaning toward writing a wrapper for tmpfile() which will use fopencookie() on Linux and fall back to the system tmpfile() everywhere else.  That would at least improve matters for Linux users, and would make things no worse on other systems.

**Comment by ghlawrence2000** [comment]
I am new to this game, but patiently persevered and managed to get freetype26 working fine in visual studio2013. My sole requirement of Freetype is to produce 1bpp monochrome bitmaps from ttf fonts, it works fine up to about 34 pixels height, after that it crashes. I am aiming to import fonts into an Arduino project if anyone is interested.

I am convinced it is a memory related problem as the process fails sooner if the requested height is bigger.

This is my code :- [code]	for (n = 33; n < 127; n++)
	{
		fprintf(stdout, "before get_char_index\n");
		int index = FT_Get_Char_Index(face, n);
		fprintf(stdout, "before load glyph\n");
		error = FT_Load_Glyph(face, index, FT_RENDER_MODE_MONO);
		if (error) {
			fprintf(stdout, "problem loading glyph!\n"); break;
		}
		fprintf(stdout, "before render glyph\n");
		error = FT_Render_Glyph(face->glyph, FT_RENDER_MODE_MONO);
		if (error) {
			fprintf(stdout, "problem rendering glyph!\n"); break;
		}
		fprintf(stdout, "after render glyph\n");
		FT_GlyphSlot slot = face->glyph;
		FT_Bitmap &b1 = slot->bitmap;
		int w = b1.width;
		int h = b1.rows;
		int widthBytes = w / 8;
		int widthRemain = w % 8;
		int bufsize;
		if (w <= 16) { bufsize = 2; }
		else if ((w > 16 && w <= 32)) { bufsize = 4; }
		else if (w > 32 && w <= 64) { bufsize = 8; }
		else if (w > 64 && w < 128) { bufsize = 16; }
		fprintf(stdout, "char %d wb %d, wr %d, x %d, y %d bytes %x topoffset %d\n", n, widthBytes, widthRemain, w, h, bufsize*h, maxBearing - (metrics[n].horiBearingY / 64));
		//unsigned char* data = new unsigned char[(bufsize*h)];
		//memset(data, 0, (bufsize * h));
		//memcpy(data, b1.buf=fer, (bufsize * h));
		fputc(w, pFile);
		int pack = 0;
		if (maxBearing - (metrics[n].horiBearingY / 64) > 0)
		{
			for (int z = 0; z < (maxBearing - (metrics[n].horiBearingY / 64)); z++)
			{
				for (int y = 0; y < bufsize; y++)
				{
					putc(0, pFile);
					pack++;
				}
			}
		}
		//fprintf(stdout, "Packbytes %d topoffset %d buffsize %d\n", pack, maxBearing - (metrics[n].horiBearingY / 64), bufsize);
		for (int q = 0; q < bufsize*h; q++)
		{
			// revbits just reverses the bits eg 0x00111010 becomes 0x01011100
			int myByte = b1.buffer[q];
			int myByterev = revBits(myByte);
			b1.buffer[q] = myByterev;
		}
		fwrite(b1.buffer, 1, (bufsize*h), pFile);
		fprintf(stdout, "after filewrite\n");

		if (h + maxBearing - (metrics[n].horiBearingY / 64) < (maxBearing - minHang))
		{
			//fprintf(stdout, "got into if\n");
			for (int z = h + maxBearing - (metrics[n].horiBearingY / 64); z < (maxBearing - minHang); z++)
			{
				for (int y = 0; y < bufsize; y++)
				{
					putc(0, pFile);
				}
				//fprintf(stdout, "z loop %d * bufsize byte ", z);
			}
			fprintf(stdout, "exit z loop\n");
		}
		//delete[] data;
	}
[/code]

This is the result where it goes wrong with a requested pixel height of 35 using Arial.ttf as source.

[quote]before render glyph
after render glyph
char 86 wb 2, wr 7, x 23, y 25 bytes 64 topoffset 1
after filewrite
exit z loop
before get_char_index
before load glyph
before render glyph
after render glyph
char 87 wb 4, wr 1, x 33, y 25 bytes c8 topoffset 1
after filewrite
exit z loop
before get_char_index
before load glyph[/quote]

This is the result where it goes wrong with a requested pixel height of 36 using Arial.ttf as source.

[quote]after render glyph
char 63 wb 2, wr 0, x 16, y 26 bytes 34 topoffset 1
after filewrite
exit z loop
before get_char_index
before load glyph
before render glyph
after render glyph
char 64 wb 4, wr 1, x 33, y 34 bytes 110 topoffset 1
after filewrite
before get_char_index
before load glyph
[/quote]

And it will continue to crash at the same point as above until a requested pixel height of 41, but with a requested pixel height of 42 it will now crash here :- 

[quote]after render glyph
char 36 wb 2, wr 3, x 19, y 37 bytes 94 topoffset 0
after filewrite
exit z loop
before get_char_index
before load glyph
before render glyph
after render glyph
char 37 wb 4, wr 1, x 33, y 32 bytes 100 topoffset 2
after filewrite
exit z loop
before get_char_index
before load glyph[/quote]

It will then continue to crash at the same point all the way up to a requested height of 61, at 62 it crashes here :-

[quote]after render glyph
char 34 wb 2, wr 0, x 16, y 15 bytes 1e topoffset 2
after filewrite
exit z loop
before get_char_index
before load glyph
before render glyph
after render glyph
char 35 wb 4, wr 1, x 33, y 46 bytes 170 topoffset 2
after filewrite
exit z loop
before get_char_index
before load glyph[/quote]

It would appear to me that the crash occurs when the character bitmap width exceeds 32 bits....

How can I increase this?

Regards,

Graham

**Comment by mskala** [comment]
FWIW, this is now fixed in FontAnvil, by following @lemzwerg's suggestion of making the temporary font be CFF instead of PS.  I don't have a clean patch that could be applied to FontForge, because the code in question in FontAnvil has been remodelled to use a FontAnvil-specific file abstraction layer and read directly out of a temporary memory buffer instead of using mmap().  However, it probably wouldn't be hard to do a simpler change in FontForge and just make it use CFF format while keeping the FontForge file-on-disk architecture.  I'm not volunteering, because my own problem is solved and the absence of any reproduction of this bug by anyone else suggests that it may not really be an issue for many FontForge users.
